sys/kern/proc.c

```diff
@@ -569,42 +569,76 @@ __noreturn void proc_exit(int exitstatus) {
   thread_exit();
 }
 
+static int proc_pgsignal(cred_t *cred, session_t *s, pgid_t pgid, signo_t sig) {
+  pgrp_t *pgrp = NULL;
+
+  WITH_MTX_LOCK (all_proc_mtx) {
```

Comment:
We can hold `all_proc_mtx` for the whole duration of the function. This can hurt performance a little bit, but it simplifies the code. It's also required after you make the changes I suggested in `proc_cansignal()`.